### ISSUE TITLE
When I replaced the model in the tutorial with my own other model, this error occurred

### ISSUE THREAD

**Issue by zxf1336899252** [tianyic/only_train_once#46]
When I replaced the model in the tutorial with my own other model, this error occurred：
This is my code：
student_model = STU_IAT().cuda()
#student_model = STU_IAT()
if config.pretrain_dir is not None:
    student_model.load_state_dict(torch.load(config.pretrain_dir))

from sanity_check.backends.resnet_cifar10 import resnet18_cifar10
#model = resnet18_cifar10().cuda()
dummy_input = torch.rand(1, 3, 400, 320)
oto = OTO(model=student_model.cuda(), dummy_input=dummy_input.cuda())
oto.visualize(view=False, out_dir='../cache')
print("@@@")
# Data Setting
train_dataset = lowlight_loader(images_path=config.img_path, normalize=config.normalize)
train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=config.batch_size, shuffle=True, num_workers=8,
                                           pin_memory=True)
val_dataset = lowlight_loader(images_path=config.img_val_path, mode='test', normalize=config.normalize)
val_loader = torch.utils.data.DataLoader(val_dataset, batch_size=1, shuffle=False, num_workers=8, pin_memory=True)

# optimizer = torch.optim.Adam(student_model.parameters(), lr=config.lr, weight_decay=config.weight_decay)

# scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=config.num_epochs)

device = next(student_model.parameters()).device
print('the device is:', device)
hard_loss = F.mse_loss
alpha = 0.2
soft_loss = F.smooth_l1_loss

ssim_high = 0
psnr_high = 0

example_inputs = torch.randn(1, 3, 400, 320).to(device)

optimizer = oto.hesso(
    variant='sgd', 
    lr=0.1, 
    weight_decay=1e-4,
    target_group_sparsity=0.7,
    start_pruning_step=10 * len(train_loader), 
    pruning_periods=10,
    pruning_steps=10 * len(train_loader)
)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=50, gamma=0.1) 

student_model.train()
print('######## Start IAT Training #########')
![image](https://github.com/tianyic/only_train_once/assets/65944463/5ee59c32-feb9-4c2b-b066-09417c77595c)
请问这是为什么？报错信息显示是oto = OTO(model=student_model.cuda(), dummy_input=dummy_input.cuda()) 这一步有问题

**Comment by tianyic** [comment]
@zxf1336899252 thanks for reaching out. 属实头一次见到这里会raise errors，maybe有一些之前没考虑过的case，可否提供一个demonet for further investigation？

**Comment by zxf1336899252** [comment]
> @zxf1336899252 thanks for reaching out. 属实头一次见到这里会raise errors，maybe有一些之前没考虑过的case，可否提供一个demonet for further investigation？

感谢作者！
这是我的模型，
import torch
import torch.nn as nn
import math
import os
from timm.models.layers import trunc_normal_
from model.xlc_global_net import Global_pred
# from model.xlc_vit_global_net import ViT

class Aff_channel(nn.Module):
    ''' 先将图片乘以color矩阵，然后计算y = ax + b '''
    def __init__(self, dim, channel_first = True):
        super().__init__()
        # learnable
        self.alpha = nn.Parameter(torch.ones([1, 1, dim]))
        self.beta = nn.Parameter(torch.zeros([1, 1, dim]))
        # 生成一个n*n的单位矩阵E，用以初始化color
        # self.color = nn.Parameter(torch.eye(dim))
        self.channel_first = channel_first

    def forward(self, x):
        if self.channel_first:
            x = x * self.alpha + self.beta
        return x

class CMlp(nn.Module):
    ''' 使用卷积模仿全连接层 '''
    def __init__(self, in_features=16, hidden_features=32, out_features=16, act_layer=nn.LeakyReLU(negative_slope=0.2, inplace=True)):
        super().__init__()
        self.fc1 = nn.Conv2d(in_features, hidden_features, 1)
        self.act1 = act_layer
        self.act2 = act_layer
        self.fc2 = nn.Conv2d(hidden_features, out_features, 1)
        self.bn1 = nn.BatchNorm2d(hidden_features)
        self.bn2 = nn.BatchNorm2d(out_features)

    def forward(self, x):
        x = self.fc1(x)
        x = self.act1(x)
        x = self.bn1(x)
        x = self.fc2(x)
        x = self.act2(x)
        x = self.bn2(x)
        return x

class CBlock_ln(nn.Module):
    ''' IAT重要组成模块 '''
    def __init__(self, dim=16, mlp_ratio=2, norm_layer=Aff_channel, act_layer=nn.LeakyReLU(negative_slope=0.2,inplace=True), init_values=1e-4):
        super().__init__()
        self.pos_embed = nn.Conv2d(dim, dim, 3, padding=1, groups=dim)
        self.norm1 = norm_layer(dim)
        self.conv1 = nn.Conv2d(dim, dim, 1)
        self.conv2 = nn.Conv2d(dim, dim, 1)
        self.conv3 = nn.Conv2d(dim, dim, 1)
        self.conv4 = nn.Conv2d(dim, dim, 1)
        self.conv5 = nn.Conv2d(dim, dim, 1, groups=dim)
        self.attn1 = nn.Conv2d(dim, dim, 3, padding=1, groups=dim)
        self.attn2 = nn.Conv2d(dim, dim, 3, padding=1, groups=dim)
        self.norm2 = norm_layer(dim)
        mlp_hidden_dim = int(dim * mlp_ratio)
        self.gamma_1 = nn.Parameter(init_values * torch.ones((1, dim, 1, 1)), requires_grad=True)
        self.gamma_2 = nn.Parameter(init_values * torch.ones((1, dim, 1, 1)), requires_grad=True)
        self.mlp = CMlp(dim, mlp_hidden_dim, dim, act_layer=act_layer)

    def forward(self, x):
        x = x + self.pos_embed(x)
        B, C, H, W = x.shape
        # B, C, H, W = int(B), int(C), int(H), int(W)
        norm_x = self.conv1(x)
        norm_x = norm_x.flatten(2).transpose(1, 2)
        norm_x = self.norm1(norm_x)
        norm_x = norm_x.reshape(-1, int(H), int(W), int(C)).permute(0, 3, 1, 2)
        norm_x = self.conv5(norm_x)
        norm_x = norm_x + self.attn1(self.conv2(norm_x))
        norm_x = norm_x + self.conv3(self.attn2(norm_x))
        x = x + self.gamma_1*norm_x
        norm_x = self.conv4(x)
        norm_x = norm_x.flatten(2).transpose(1, 2)
        norm_x = self.norm2(norm_x)
        norm_x = norm_x.reshape(-1, int(H), int(W), int(C)).permute(0, 3, 1, 2)
        norm_x = self.conv5(norm_x)
        x = x + self.gamma_2*(norm_x + self.mlp(norm_x))

        return x




class Local_pred(torch.nn.Module):
    ''' 生成mul矩阵和add矩阵，y=ax + b  '''
    def __init__(self, in_dim=3, dim=16):
        super(Local_pred, self).__init__()
        self.conv1 = nn.Conv2d(in_dim, dim//2, 3, padding=1)
        
        self.conv2 = nn.Conv2d(dim//2, dim,1, groups=dim//2)
        conv_mul = nn.Conv2d(dim, 3, 3, padding=1)
        conv_add = nn.Conv2d(dim, 3, 3, padding=1)
        self.relu = torch.nn.LeakyReLU(negative_slope=0.3, inplace=True)
        self.bn1 = nn.BatchNorm2d(dim)
        # block1 = [CBlock_ln(dim), CBlock_ln(dim, mlp_ratio=2)]# , CBlock_ln(16, mlp_ratio=2)]# , CBlock_ln(16, mlp_ratio=4)]
        block2 = [CBlock_ln(dim), CBlock_ln(dim, mlp_ratio=2)]# , CBlock_ln(16, mlp_ratio=2)]# , CBlock_ln(16, mlp_ratio=4)]
        # self.mul_blocks = nn.Sequential(*block1, conv_mul, nn.ReLU(), nn.BatchNorm2d(3))
        self.add_blocks = nn.Sequential(*block2)
        self.add_end = nn.Sequential(conv_add, nn.LeakyReLU(negative_slope=0.2,inplace=True), nn.BatchNorm2d(3))
    def forward(self, x):
        x = self.bn1(self.relu(self.conv2(self.conv1(x))))
        # mul = self.mul_blocks(x)
        add = self.add_end(self.add_blocks(x))

        # return mul, add
        return add


class xlc_IAT(nn.Module):
    '''  IAT student 模型  '''
    def __init__(self, in_dim=3):
        super(xlc_IAT, self).__init__()
        self.local_net = Local_pred(in_dim)
        self.global_net = Global_pred(in_channels=in_dim)
        # self.vit_model = ViT(
        #     image_size=(400, 600),
        #     patch_size=(4, 6),
        #     num_classes=10,
        #     dim=32,
        #     depth=1,
        #     heads=2,
        #     mlp_dim=64,
        #     dropout=0.1,
        #     emb_dropout=0.1
        # )

    def apply_color(self, image, ccm):
        shape = image.shape
        image = image.view(-1, 3)
        image = torch.tensordot(image, ccm, dims=[[-1], [-1]])
        image = image.view(shape)
        return torch.clamp(image, 1e-8, 1.0)

    def forward(self, x):
        # mul, add = self.local_net(x)
        add = self.local_net(x)
        # img_high = (x.mul(mul)).add(add)
        img_high = x.add(add)
        gamma, color = self.global_net(x)
        # gamma_color = self.vit_model(x)
        # gamma, color = gamma_color[:, :1], gamma_color[:, 1:].view(-1, 3, 3)
        b = img_high.shape[0]
        img_high = img_high.permute(0, 2, 3, 1)  # (B,C,H,W) -- (B,H,W,C)
        img_high = torch.stack([self.apply_color(img_high[i,:,:,:], color[i,:,:])**gamma[i,:] for i in range(b)], dim=0)

        img_high = img_high.permute(0, 3, 1, 2)  # (B,H,W,C) -- (B,C,H,W)
        # print(img_high.shape)
        return img_high



大致结构是这样的， class xlc_IAT是我模型的头，请问是有哪些层不支持吗？我还需要提供什么信息才能更好的定位我的问题

**Comment by zxf1336899252** [comment]
> @zxf1336899252 thanks for reaching out. 属实头一次见到这里会raise errors，maybe有一些之前没考虑过的case，可否提供一个demonet for further investigation？
感谢作者！
这是我的模型，
![image](https://github.com/tianyic/only_train_once/assets/65944463/95d58cf6-7ea3-4592-956d-0820f5b05e3d)
![image](https://github.com/tianyic/only_train_once/assets/65944463/351b6ab3-ff55-4c1d-88c1-c412a424fc5d)
![image](https://github.com/tianyic/only_train_once/assets/65944463/544bd821-eacf-48bb-86f3-e8af267cf419)
![image](https://github.com/tianyic/only_train_once/assets/65944463/2873dd60-842a-47b3-be6f-79f164c6d935)

请问是我模型中有些层不支持吗？ 或者还需要什么信息才能更好的分析这个问题，万分感谢

**Comment by tianyic** [comment]
这个网络看上去总体没问题，应该可以train & prune，但是有几个trainable nn.Parameters 看起来会是造成错误的root causes。现在的framework对于没有包含在supported modules里的trainable nn.Parameters会有一定麻烦。解决起来需要到[Operator](https://github.com/tianyic/only_train_once/blob/cbb3d3dccf95c383e9cddcbaf8592cf3db13817b/only_train_once/operation/operator.py#L646) 里添加一个AFF_channel，需要一定工程量。
